server/blob/blob-joining/src/main/java/org/apache/james/blob/joining/JoiningBlobStore.java

```diff
@@ -26,52 +26,133 @@
+    }
+
+    private final JoiningBlobStoreFallBack fallBackBlobStore;
     private final BlobStore primaryBlobStore;
     private final BlobStore secondaryBlobStore;
 
     @VisibleForTesting
     JoiningBlobStore(BlobStore primaryBlobStore, BlobStore secondaryBlobStore) {
         this.primaryBlobStore = primaryBlobStore;
         this.secondaryBlobStore = secondaryBlobStore;
+        this.fallBackBlobStore = new JoiningBlobStoreFallBack(primaryBlobStore, secondaryBlobStore);
     }
 
     @Override
     public CompletableFuture<BlobId> save(byte[] data) {
-        return primaryBlobStore.save(data);
+        try {
+            return fallBackBlobStore.save(data);
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while saving bytes data, fall back to secondary blob store", e);
+            return secondaryBlobStore.save(data);
+        }
     }
 
     @Override
     public CompletableFuture<BlobId> save(InputStream data) {
-        return primaryBlobStore.save(data);
+        try {
+            return fallBackBlobStore.save(data);
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while saving InputStream data, fall back to secondary blob store", e);
+            return secondaryBlobStore.save(data);
+        }
     }
 
     @Override
     public CompletableFuture<byte[]> readBytes(BlobId blobId) {
-        return primaryBlobStore.readBytes(blobId)
-            .thenCompose(primaryResultInBytes -> readFromSecondaryIfNeeded(blobId, primaryResultInBytes));
+        try {
+            return fallBackBlobStore.readBytes(blobId);
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while readBytes, fall back to secondary blob store", e);
+            return secondaryBlobStore.readBytes(blobId);
+        }
     }
 
     @Override
     public InputStream read(BlobId blobId) {
-        return Optional.ofNullable(primaryBlobStore.read(blobId))
-            .filter(Throwing.<InputStream>predicate(inputStream -> inputStream.available() > 0).sneakyThrow())
-            .orElseGet(() -> secondaryBlobStore.read(blobId));
+        try {
+            return Optional.ofNullable(primaryBlobStore.read(blobId))
+                .filter(Throwing.<InputStream>predicate(inputStream -> inputStream.available() > 0).sneakyThrow())
+                .orElseGet(() -> secondaryBlobStore.read(blobId));
+        } catch (Exception e) {
+            LOGGER.error("exception directly happens while read, fall back to secondary blob store", e);
+            return secondaryBlobStore.read(blobId);
+        }
     }
 
-    private CompletionStage<byte[]> readFromSecondaryIfNeeded(BlobId blobId, byte[] primaryResultInBytes) {
-        if (isNullOrEmpty(primaryResultInBytes)) {
-            return secondaryBlobStore.readBytes(blobId);
-        }
-        return CompletableFuture.completedFuture(primaryResultInBytes);
+    @VisibleForTesting
+    BlobStore getPrimaryBlobStore() {
```

Comment:
Can we avoid these getters?

Exposing the internal of this class is not needed for your tests as you can have fields holding this information (you have the info in your tests, so no need to read the info from the joining blob store)